packages/esm-active-visits-app/src/active-visits-widget/active-visits.resource.tsx

```diff
@@ -142,6 +145,174 @@ export function useActiveVisits() {
   };
 }
 
+export function useObsConcepts(uuids: Array<string>): {
+  obsConcepts: Array<OpenmrsResource> | undefined;
+  isLoadingObsConcepts: boolean;
+} {
+  const fetchConcept = async (uuid: string): Promise<OpenmrsResource | null> => {
+    try {
+      const response = await openmrsFetch(`${restBaseUrl}/concept/${uuid}?v=custom:(uuid,display)`);
+      return response?.data;
+    } catch (error) {
+      console.error(`Error fetching concept for UUID: ${uuid}`, error);
+      return null;
+    }
+  };
+
+  const { data, isLoading, error } = useSWR(uuids.length > 0 ? ['obs-concepts', uuids] : null, async () => {
+    const results = await Promise.all(uuids.map(fetchConcept));
+    return results.filter((concept) => concept !== null);
+  });
+
+  return useMemo(
+    () => ({
+      obsConcepts: data ?? [],
+      isLoadingObsConcepts: isLoading,
+    }),
+    [data, isLoading],
+  );
+}
+
+export function useActiveVisitsSorting(tableRows: Array<any>) {
+  const [sortParams, setSortParams] = useState<{
+    key: string;
+    sortDirection: 'ASC' | 'DESC' | 'NONE';
+  }>({ key: 'visitStartTime', sortDirection: 'DESC' });
+
+  const sortRow = (cellA, cellB, { key, sortDirection }) => {
+    setSortParams({ key, sortDirection });
+  };
+
+  const getSortValue = (item: any, key: string) => {
+    // For observation columns
+    if (key.startsWith('obs-')) {
+      const conceptUuid = key.replace('obs-', '');
+      const obsValue = item?.observations?.[conceptUuid]?.[0]?.value;
+
+      if (!obsValue) return null;
+      if (typeof obsValue === 'object' && obsValue.display) {
+        return obsValue.display.toLowerCase();
+      }
+      return obsValue;
+    }
+
+    const value = item[key];
+    if (value == null) return null;
+
+    if (key === 'visitStartTime') {
+      return new Date(value).getTime();
+    }
+
+    if (key === 'age' && !isNaN(value)) {
+      return Number(value);
+    }
+
+    return String(value).toLowerCase();
+  };
+
+  const sortedRows = useMemo(() => {
+    if (sortParams.sortDirection === 'NONE') {
+      return tableRows;
+    }
+
+    return [...tableRows].sort((a, b) => {
+      const valueA = getSortValue(a, sortParams.key);
+      const valueB = getSortValue(b, sortParams.key);
+
+      if (valueA === null && valueB === null) return 0;
+      if (valueA === null) return 1;
+      if (valueB === null) return -1;
+
+      if (typeof valueA === 'number' && typeof valueB === 'number') {
+        return sortParams.sortDirection === 'DESC' ? valueB - valueA : valueA - valueB;
+      }
+
+      const compareResult = String(valueA).localeCompare(String(valueB), undefined, {
+        numeric: true,
+      });
+
+      return sortParams.sortDirection === 'DESC' ? -compareResult : compareResult;
+    });
+  }, [sortParams, tableRows]);
+
+  return {
+    sortedRows,
+    sortRow,
+  };
+}
+
+export function useTableHeaders(t, config, obsConcepts) {
```

Comment:
Since this is a hook, you can directly use the `useTranslation` and `useConfig` in this hook.